agent/arty_uploader.go

```
@@ -0,0 +1,168 @@
+	password := os.Getenv("BUILDKITE_ARTIFACTORY_PASSWORD")
+	if stringURL == "" || username == "" || password == "" {
+		return nil, errors.New("Must set BUILDKITE_ARTIFACTORY_URL, BUILDKITE_ARTIFACTORY_USER, BUILDKITE_ARTIFACTORY_PASSWORD when using rt:// path")
+	}
+	parsedURL, err := url.Parse(stringURL)
+	if err != nil {
+		return nil, err
+	}
+	return &ArtifactoryUploader{
+		logger:     l,
+		conf:       c,
+		client:     &http.Client{},
+		iURL:       parsedURL,
+		Path:       path,
+		Repository: repo,
+		user:       username,
+		password:   password,
+	}, nil
+}
+
+func ParseArtifactoryDestination(destination string) (repo string, path string) {
+	parts := strings.Split(strings.TrimPrefix(string(destination), "rt://"), "/")
+	path = strings.Join(parts[1:len(parts)], "/")
+	repo = parts[0]
+	return
+}
+
+func (u *ArtifactoryUploader) URL(artifact *api.Artifact) string {
+	url := u.iURL
+	// ensure proper URL formatting for upload
+	url.Path = strings.Join([]string{
+		strings.Trim(url.Path, "/"),
+		u.artifactPath(artifact),
+	}, "/")
+	return url.String()
+}
+
+func (u *ArtifactoryUploader) Upload(artifact *api.Artifact) error {
+	// Open file from filesystem
+	u.logger.Debug("Reading file \"%s\"", artifact.AbsolutePath)
+	f, err := os.Open(artifact.AbsolutePath)
+	if err != nil {
+		return fmt.Errorf("failed to open file %q (%v)", artifact.AbsolutePath, err)
+	}
+
+	// Upload the file to Artifactory.
+	u.logger.Debug("Uploading \"%s\" to `%s`", artifact.Path, u.Repository)
+
+	req, err := http.NewRequest("PUT", u.iURL.String(), f)
+	req.SetBasicAuth(u.user, u.password)
+	if err != nil {
+		return err
+	}
+
+	res, err := u.client.Do(req)
+	if err != nil {
+		return err
+	}
+	if err := checkResponse(res); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (u *ArtifactoryUploader) artifactPath(artifact *api.Artifact) string {
+	jobID := os.Getenv("BUILDKITE_JOB_ID")
+	if jobID == "" {
+		jobID = "NO-JOB-ID"
```

Might be worth resolving this in the constructor so you can return an error?

---

I figured the only time this would be true would be in a develpment space -- would say it's good to have all those variables in one place :) Can change if you don't agree!

---

updated with a warning separate from authentication warning for JOB ID -- want to avoid confusion for anyone using pipelines / using artifact upload for local development.